server/src/main/java/io/druid/segment/realtime/firehose/EventReceiverFirehoseFactory.java

```diff
@@ -393,5 +402,68 @@ public boolean isClosed()
     {
       return closed;
     }
+
+    /**
+     * Checks the request for a producer ID and sequence value.  If the producer ID is specified, a corresponding
+     * sequence value must be specified as well.  If the incoming sequence is less than or equal to the last seen
+     * sequence for that producer ID, the request is ignored
+     *
+     * @param req Http request
+     * @param responseContentType Response content type
+     * @param responseMapper Response object mapper
+     * @return Optional of a response to return of an empty optional if the request can proceed
+     */
+    private Optional<Response> checkProducerSequence(
+        final HttpServletRequest req,
+        final String responseContentType,
+        final ObjectMapper responseMapper
+    )
+    {
+      final String producerId = req.getHeader("X-Firehose-Producer-Id");
+
+      if (producerId == null) {
+        return Optional.empty();
+      }
+
+      final String sequenceValue = req.getHeader("X-Firehose-Producer-Seq");
+
+      if (sequenceValue == null) {
+        return Optional.of(
+            Response.status(Response.Status.BAD_REQUEST)
+                       .entity(ImmutableMap.<String, Object>of("error", "Producer sequence value is missing"))
+                       .build()
+        );
+      }
+
+      Long producerSequence = producerSequences.computeIfAbsent(producerId, key -> Long.MIN_VALUE);
+
+      try {
+        Long newSequence = Long.parseLong(sequenceValue);
+        if (newSequence <= producerSequence) {
+          return Optional.of(
+              Response.ok(
+                responseMapper.writeValueAsString(
+                    ImmutableMap.of("eventCount", 0, "skipped", true)
+                ),
+                responseContentType
+            ).build()
+          );
+        }
+
+        producerSequences.put(producerId, newSequence);
```

Comment:
typically in this kind of design, I would expect sequence numbers to also be sequential i.e. always incrementing by one so that there is no possibility of missing a batch either.

Reply:
Thanks @himanshug.  I'm open to that.  My only concern with that is that it forces the consumers to maintain that state on their end. 

With the design as is, the user _could_ maintain a sequential counter but they could also use something like the wall time millis (assuming they aren't sending more than one 1000/second) and not maintain any state at all.

Thoughts?

Reply:
i think given that the focus of this PR is someidempotency rather than exactly once, so its fine to keep it because providing exactly once does require client to maintain sequence in a persistent store so that it could start with correct sequenceId on restart too which might be decent overhead. for exactly once semantics , i think users should probably use kafka-indexing-service.
however, note that with current design there is possibility of drop/duplicate of last batch on client crash and restart.